src/Analyzers/Core/Analyzers/UseConditionalExpression/ForAssignment/UseConditionalExpressionForAssignmentHelpers.cs

```diff
@@ -49,9 +49,40 @@ internal static class UseConditionalExpressionForAssignmentHelpers
                 return false;
             }
 
+            // If both assignments are discards type check is required.
+            // Since discard can discard value of any type, after converting if statement to a conditional expression
+            // it can produce compiler error if types are not the same and there is no implicit conversion between them, e.g.:
+            // if (flag)
+            // {
+            //     _ = 5;
+            // }
+            // else
+            // {
+            //     _ = "";
+            // }
+            // This code can result in `_ = flag ? 5 : ""`, which immediately produces CS0173
+            if (trueAssignment?.Target is IDiscardOperation &&
+                falseAssignment?.Target is IDiscardOperation &&
+                !AreEqualOrHaveImplicitConversion(trueAssignment.Type, falseAssignment.Type, trueAssignment.SemanticModel!.Compilation))
+            {
+                return false;
+            }
+
             isRef = trueAssignment?.IsRef == true;
             return UseConditionalExpressionHelpers.CanConvert(
                 syntaxFacts, ifOperation, trueStatement, falseStatement);
+
+            static bool AreEqualOrHaveImplicitConversion(ITypeSymbol? firstType, ITypeSymbol? secondType, Compilation compilation)
+            {
+                if (firstType?.Equals(secondType, SymbolEqualityComparer.Default) == true)
```

Comment:
```suggestion
                if (SymbolEqualityComparer.Default.Equals(firstType, secondType))
```

Note: This is not a stylistic-only suggestion. If both types are null, previously we would be returning `false`, but now we will be returning `true`. I don't know if both can be null or not, but just for robustness. Also, this suggestion is quite shorter :)